pyrealm/phenology/fapar_limitation.py

```diff
@@ -0,0 +1,274 @@
+        of all values of the year)
+    """
+
+    # Extract years from datetimes
+    all_years = datetimes.astype("datetime64[Y]")
+
+    if len(x) == len(growing_season):  # this is daily data
+        daily_x = x
+        n_days = len(x)
+        years_by_day = all_years.view()
+        obs_per_day = int(len(datetimes) / n_days)
+        years_by_day.shape = tuple([n_days, obs_per_day, *list(all_years.shape[1:])])
+    elif len(x) == len(datetimes):  # this is subdaily data
+        # Create scaler object to handle conversion between scales
+        scaler = SubdailyScaler(datetimes)
+        scaler.set_nearest(np.timedelta64(12, "h"))
+        # Convert values to daily to match with growing_season
+        daily_x = scaler.get_daily_means(x)
+        years_by_day = scaler.get_window_values(np.asarray(all_years))
+    else:
+        raise ValueError("Input array does not fit datetimes nor growing_season array")
+
+    # Which years are present?
+    years = np.unique(all_years)
+
+    # Compute annual totals or means, only taking into account the days which are in
+    # growing season.
+
+    annual_x = np.zeros(len(years))
+
+    if method == "total":
+        for i in range(len(years)):
+            annual_x[i] = sum(
+                daily_x[growing_season & (years_by_day[:, 0] == years[i])].astype(
+                    np.int64
+                )
+            )
+    elif method == "mean":
+        for i in range(len(years)):
+            annual_x[i] = np.mean(
+                daily_x[growing_season & (years_by_day[:, 0] == years[i])].astype(
+                    np.int64
+                )
+            )
+    else:
+        raise ValueError("No valid method given for annual values")
+
+    return annual_x
+
+
+def compute_annual_total_precip(
+    precip: NDArray, datetimes: NDArray[np.datetime64], growing_season: NDArray[np.bool]
+) -> NDArray:
+    """Returns the sum of annual precipitation."""
+
+    annual_precip = get_annual(precip, datetimes, growing_season, "total")
+
+    return convert_precipitation_to_molar(annual_precip)
+
+
+def convert_precipitation_to_molar(precip_mm: NDArray) -> NDArray:
+    """Convert precipitation from mm/m2 to mol/m2.
+
+    - 1 mm/m2 = 1000000 mm3 = 1000 mL = 1 L
+    - Molar mass of water = 18g
+    - Assuming density = 1 (but really temp varying), molar volume = 18mL
+    - So 1 mm/m2 = 1000 / 18 ~ 55.55 mols/m2
+    """
+
+    water_mm_to_mol = 1000 / 18
+    precip_molar = precip_mm * water_mm_to_mol
+
+    return precip_molar
+
+
+class FaparLimitation:
+    r"""FaparLimitation class to compute fAPAR_max and LAI.
+
+    This class takes the annual total potential GPP and precipitation, the annual
+    mean CA, Chi and VPD, as well as the aridity index and some constants to compute
+    the annual peak fractional absorbed photosynthetically active radiation (
+    fAPAR_max) and annual peak Leaf Area Index (LAI).
```

Comment:
It would be good to add a `.. TODO::` section in this docstring just to highlight to early users where the rough edges are:

* daily or subdaily timescales at present, not monthly etc
* daily conditions taken as noon values - might need to relax this.
* users currently need to provide own growing season definition in `from_pmodel`.

Reply:
Just to make sure: do you want those as `TODO:` or as `.. todo::` or as `.. todolist::` ?

Reply:
I meant as a `sphinx` `.. TODO::` directive (https://www.sphinx-doc.org/en/master/usage/extensions/todo.html) as it does nice styling.

Although - not sure I've configured that extension in this repo!

Reply:
@MarionBWeinzierl Are you happy for me to push a commit to this turning on that extension and styling it?

Reply:
Yes, of course. I had searched in the repo whether, but I couldn't find any other place where the sphinx directive was used, that's why I was unsure whether that's the right one to use here.